spark3/main.py

```diff
@@ -75,14 +76,22 @@ def parse_trace_to_function(df: DataFrame,
 
         df.sql_ctx.udf.registerJavaFunction("decode_func_%s" % name, DECODE_CONTRACT_FUNCTION_UDF, schema)
 
-        if len([col_name for col_name, col_type in df.dtypes if
-                col_name == "input" and col_type == "string"]) != 1:
-            raise ColumnNotFoundInDataFrame("input", df)
+        if not contains_column(df.dtypes, "input", "string"):
+            raise ColumnNotFoundInDataFrame("input(string)", df)
+
+        if not contains_column(df.dtypes, "unhex_input", "binary"):
+            # len is required argument in substring, use expr(substring) to replace
+            df = df.withColumn("unhex_input", unhex(expr('substring(input, 3)')))
+
+        if contains_column(df.dtypes, "output", "string"):
+            if not contains_column(df.dtypes, "unhex_output", "binary"):
+                # len is required argument in substring, use expr(substring) to replace
+                df = df.withColumn("unhex_output", unhex(expr('substring(output, 3)')))
```

Comment:
The assumption of data format is strong here.

If user's data has no 0x prefix, this assumption will make him conduct string concat before hands.